### PR TITLE
first pass on merging eslint rules

### PR DESCRIPTION
**Pull request by rsun19**
Partially Closes: [RHOAIENG-2534](https://issues.redhat.com/browse/RHOAIENG-2534)

## Description

Added frontend `eslint`  and `tsconfig` rules to the backend, as well as `package.json` packages.

You can see the added rules [here](https://github.com/rsun19/odh-dashboard/commit/104e1f22f2d53f332d2d897718572f515d2f869a)

## How Has This Been Tested?

`cd backend && npm run test`
`npm run build`

## Test Impact

I have not implemented functionality, just refactored code.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work (`ran npm run test`)
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

File: backend/src/utils/route-security.ts [hunk 1]
```diff
@@ -96,7 +96,7 @@ const requestSecurityGuard = async (
   }
 
   // Api with no name object, allow reads
-  if (name == null && namespace === notebookNamespace && isReadRequest) {
+  if (!name && namespace === notebookNamespace && isReadRequest) {
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/utils/route-security.ts:99]: should i delete?

[the remark this answers is not in the document]
Comment (christianvogt) [reply backend/src/utils/route-security.ts:99]: You should change the name parameter to be optional instead. You will need to change the callers back to no longer provide a string with `?? ''`.

File: backend/src/utils/route-security.ts [hunk 2]
```diff
@@ -161,23 +161,22 @@ const handleSecurityOnRouteData = async (
     await requestSecurityGuard(
       fastify,
       request,
-      request.body.metadata.name,
-      request.body.metadata.namespace,
+      request.body.metadata?.name ?? '',
+      request.body.metadata?.namespace ?? '',
```
Comment (christianvogt) [inline backend/src/utils/route-security.ts:165]: Undo this change.

File: backend/src/utils/route-security.ts [hunk 3]
```diff
@@ -161,23 +161,22 @@ const handleSecurityOnRouteData = async (
...
       needsAdmin,
     );
   } else if (isRequestParams(request)) {
     await requestSecurityGuard(
       fastify,
       request,
-      request.params.name,
+      request.params.name ?? '',
```
Comment (christianvogt) [inline backend/src/utils/route-security.ts:172]: Undo this change.

File: backend/src/utils/route-security.ts (hunk 2 above)
Comment (rsun19) [reply backend/src/utils/route-security.ts:165]: 'request.body.metadata' is possibly 'undefined'

from this type definition in `types.ts`

```
export type K8sResourceCommon = {
  metadata?: {
    name?: string;
    namespace?: string;
    generateName?: string;
    uid?: string;
    labels?: { [key: string]: string };
    annotations?: { [key: string]: string };
    creationTimestamp?: Date;
  };
} & K8sResourceBase;
```

File: backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts [hunk 4]
```diff
@@ -50,8 +66,11 @@ export const deleteAcceleratorProfile = async (
   fastify: KubeFastifyInstance,
   request: FastifyRequest,
 ): Promise<{ success: boolean; error: string }> => {
-  const customObjectsApi = fastify.kube.customObjectsApi;
-  const namespace = fastify.kube.namespace;
+  if (typeof fastify.kube === 'undefined') {
+    throw new Error('fastify.kube is not defined.');
+  }
+  const { customObjectsApi } = fastify.kube;
+  const { namespace } = fastify.kube;
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts:73]: One problem is that lots of functions take in `FastifyInstance` as the parameter like: `export default fp(async (fastify: FastifyInstance) => {`. However, when it tries to call a function requiring `KubeFastifyInstance`, there's a type mismatch now that `kube` is required. How should this be dealt with? A typeguard essentially checking if `kube` is present like `kube in fastify`?

Error: 
```
src/plugins/kube.ts:81:30 - error TS2345: Argument of type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>' is not assignable to parameter of type 'KubeFastifyInstance'.
  Property 'kube' is missing in type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>' but required in type '{ kube: KubeDecorator; }'.

81   initializeWatchedResources(fastify);
```

File: backend/src/plugins/kube.ts [hunk 5]
```diff
@@ -30,6 +31,9 @@ export default fp(async (fastify: FastifyInstance) => {
 
   let currentToken;
   try {
+    if (currentUser === null) {
+      throw new Error('Current user is null');
+    }
     currentToken = await getCurrentToken(currentUser);
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/plugins/kube.ts]: I removed the parameter instead, is this fine?

File: backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts]: Ok, i fixed this by changing the parameter types. Don't know if that's how it should be done though.

File: backend/src/utils/resourceWatcher.ts [hunk 6]
```diff
@@ -51,13 +59,13 @@ export class ResourceWatcher<T> {
...
       if (activeWatchInterval !== this.activeWatchInterval) {
-        this.activeWatchInterval = activeWatchInterval;
+        this.activeWatchInterval = activeWatchInterval ?? 0;
         if (this.activelyWatching) {
           updated = true;
         }
       }
       if (inactiveWatchInterval !== this.inactiveWatchInterval) {
-        this.inactiveWatchInterval = inactiveWatchInterval;
+        this.inactiveWatchInterval = inactiveWatchInterval ?? 0;
```
Comment (christianvogt) [inline backend/src/utils/resourceWatcher.ts:68]: These values should always be numbers.

File: backend/src/utils/userUtils.ts [hunk 7]
```diff
@@ -40,15 +40,18 @@ export const getOpenshiftUser = async (
...
 ): Promise<OpenShiftUser> => {
   try {
-    const userResponse = await fastify.kube.customObjectsApi.getClusterCustomObject(
-      'user.openshift.io',
-      'v1',
-      'users',
-      username,
-    );
-    return userResponse.body as OpenShiftUser;
+    if (fastify.kube) {
```
Comment (christianvogt) [inline backend/src/utils/userUtils.ts:43]: Unnecessary if condition.

File: backend/src/routes/api/status/adminAllowedUsers.ts [hunk 8]
```diff
@@ -44,11 +43,14 @@ const getUserActivityFromNotebook = async (
...
-        notebook.metadata?.annotations?.['kubeflow-resource-stopped'],
+      notebook.metadata.annotations['opendatahub.io/username'],
+      notebook.metadata.annotations['notebooks.kubeflow.org/last-activity'] ||
+        notebook.metadata.annotations['kubeflow-resource-stopped'],
     ])
-    .filter(([username, lastActivity]) => username && lastActivity)
+    .filter(
+      (arr): arr is [string, string] =>
+        Array.isArray(arr) && typeof arr[0] === 'string' && typeof arr[1] === 'string',
+    )
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/status/adminAllowedUsers.ts:53]: The issue is that the .map takes in undefined

```
.map<[string | undefined, string | undefined]>((notebook) => [
      notebook.metadata.annotations['opendatahub.io/username'],
      notebook.metadata.annotations['notebooks.kubeflow.org/last-activity'] ||
        notebook.metadata.annotations['kubeflow-resource-stopped'],
    ])
```

File: backend/src/routes/api/status/adminAllowedUsers.ts [hunk 9]
```diff
@@ -44,11 +43,14 @@ const getUserActivityFromNotebook = async (
 
   return notebooks.items
     .map<[string | undefined, string | undefined]>((notebook) => [
-      notebook.metadata?.annotations?.['opendatahub.io/username'],
-      notebook.metadata?.annotations?.['notebooks.kubeflow.org/last-activity'] ||
-        notebook.metadata?.annotations?.['kubeflow-resource-stopped'],
+      notebook.metadata.annotations['opendatahub.io/username'],
+      notebook.metadata.annotations['notebooks.kubeflow.org/last-activity'] ||
+        notebook.metadata.annotations['kubeflow-resource-stopped'],
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/status/adminAllowedUsers.ts:48]: I think I accounted for this with the `string | undefined` type? I'm not sure what else should be changed.

File: backend/src/routes/api/config/index.ts [hunk 10]
```diff
@@ -15,7 +15,14 @@ module.exports = async (fastify: KubeFastifyInstance) => {
...
     secureAdminRoute(fastify)(async (request: FastifyRequest, reply: FastifyReply) => {
-      reply.send(setDashboardConfig(fastify, request.body));
+      if (
+        typeof request === 'object' &&
+        'body' in request &&
+        request.body &&
+        typeof request.body === 'object'
+      ) {
+        reply.send(setDashboardConfig(fastify, request.body));
+      }
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/config/index.ts:25]: Without it, it errors:. I presume it's because it isn't defined in the `FastifyRequest` interface

```
Argument of type 'unknown' is not assignable to parameter of type 'RecursivePartial<DashboardConfig>'.ts(2345)
```

File: backend/src/routes/api/status/adminAllowedUsers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply backend/src/routes/api/status/adminAllowedUsers.ts:48]: It's this definition that's incorrect because `annotations` is defined as an object and not possibly undefined here. What happens if we change this to `annotations?: Partial<{`. Is there a large cascading effect? 
```
export type Notebook = K8sResourceCommon & {
  metadata: {
    annotations: Partial<{
      'kubeflow-resource-stopped': string; // datestamp of stop (if omitted, it is running)
      'notebooks.kubeflow.org/last-activity': string; // datestamp of last use
      'opendatahub.io/username': string; // the untranslated username behind the notebook

      // TODO: Can we get this from the data in the Notebook??
      'notebooks.opendatahub.io/last-image-selection': string; // the last image they selected
      'notebooks.opendatahub.io/last-size-selection': string; // the last notebook size they selected
    }>;
```

File: backend/src/routes/api/config/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply backend/src/routes/api/config/index.ts:25]: Ok i can understand with `request.body` having type `unknown` but why would `request` possibly be `null` here?

File: backend/src/routes/api/status/adminAllowedUsers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/status/adminAllowedUsers.ts:48]: On first glance there doesn't seem like there is a cascading effect

File: backend/src/routes/api/status/adminAllowedUsers.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/routes/api/status/adminAllowedUsers.ts:53]: Since we have `.map<[string | undefined, string | undefined]>`, shouldn't we need to filter out the `undefined` strings?

File: backend/src/routes/api/status/adminAllowedUsers.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply backend/src/routes/api/status/adminAllowedUsers.ts:48]: Careful using labels and annotations -- I caught this at random, so forgive if this is inaccurate. I do not believe we can ever assume Notebooks have the annotation or label blocks. If our types are wrong, we should fix those.

[the remark this answers is not in the document]
Comment (andrewballantyne) [reply backend/src/routes/api/status/adminAllowedUsers.ts:48]: Oh, didn't read the thread -- Christian caught this 🙂

File: backend/src/utils/userUtils.ts [hunk 11]
```diff
@@ -66,48 +69,76 @@ export const getUser = async (
...
+        return (currentUser.username || currentUser.name).split('/')[0];
+      }
+      if (
+        typeof e === 'object' &&
+        e !== null &&
+        'message' in e &&
+        typeof e.message === 'string' &&
+        'statusCode' in e &&
+        typeof e.statusCode === 'number'
+      ) {
```
[the remark this answers is not in the document]
Comment (rsun19) [reply backend/src/utils/userUtils.ts]: Created a new method instead for this instance, since `Cannot find name 'createError'.ts(2304)`

File: backend/src/utils/userUtils.ts [hunk 12]
```diff
@@ -98,16 +102,20 @@ export const getUserName = async (
...
+      fastify.log.error(`Failed to retrieve username: ${errorHandler(e)}`);
+      const error = createCustomError(
+        'Unauthorized',
+        `Failed to retrieve username: ${errorHandler(e)}`,
+        e.statusCode || 500,
+      );
+      throw error;
+    } else {
+      throw new Error(`Error getting Oauth Info for user, ${errorHandler(e)}`);
     }
```
Comment (christianvogt) [inline backend/src/utils/userUtils.ts:119]:
```suggestion
    fastify.log.error(`Failed to retrieve username: ${errorHandler(e)}`);
    const error = createCustomError(
      'Unauthorized',
      `Failed to retrieve username: ${errorHandler(e)}`,
      (isHttpError(e) && e.statusCode) || 500,
    );
    throw error;
```

File: backend/src/utils/userUtils.ts [hunk 13]
```diff
@@ -77,12 +77,16 @@ export const getUser = async (
...
+    if (isHttpError(e)) {
+      const error = createCustomError(
+        e.message,
+        `Error getting Oauth Info for user, ${errorHandler(e)}`,
+        e.statusCode,
+      );
+      throw error;
+    } else {
+      throw new Error(`Error getting Oauth Info for user, ${errorHandler(e)}`);
+    }
```
Comment (christianvogt) [inline backend/src/utils/userUtils.ts:89]: Can probably default to status code 401:
```suggestion
    const error = createCustomError(
      e.message,
      `Error getting Oauth Info for user, ${errorHandler(e)}`,
      (isHttpError(e) && e.statusCode) || 401,
    );
    throw error;
```

Comment (rsun19) [reply backend/src/utils/userUtils.ts:89]: replaced `e.message` with `errorHandler(e)`, since `e` might be undefined

File: backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts [hunk 14]
```diff
@@ -36,11 +37,11 @@ export const postAcceleratorProfile = async (
       .catch((e) => {
         throw createError(e.statusCode, e?.body?.message);
       });
-    return { success: true, error: null };
+    return { success: true, error: '' };
   } catch (e) {
-    if (e.response?.statusCode !== 404) {
+    if (createError.isHttpError(e) && e.statusCode !== 404) {
```
Comment (christianvogt) [inline backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts:42]:
```suggestion
    if (!createError.isHttpError(e) || e.statusCode !== 404) {
```

File: backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts [hunk 15]
```diff
@@ -139,11 +140,11 @@ export const updateAcceleratorProfile = async (
...
```
Comment (christianvogt) [inline backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts:145]:
```suggestion
    if (!createError.isHttpError(e) || e.statusCode !== 404) {
```

File: backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts [hunk 16]
```diff
@@ -106,17 +108,17 @@ export const updateClusterSettings = async (
...
     }
     await rolloutDeployment(fastify, namespace, 'notebook-controller-deployment');
-    return { success: true, error: null };
+    return { success: true, error: '' };
   } catch (e) {
-    fastify.log.error(e, 'Setting cluster settings error: ' + e.response?.body?.message);
-    if (e.response?.statusCode !== 404) {
-      return { success: false, error: 'Unable to update cluster settings. ' + e.message };
+    fastify.log.error(e, `Setting cluster settings error: ${errorHandler(e)}`);
+    if (isHttpError(e) && e.response.statusCode !== 404) {
```
Comment (christianvogt) [inline backend/src/routes/api/cluster-settings/clusterSettingsUtils.ts:120]:
```suggestion
    if (!isHttpError(e) || e.response.statusCode !== 404) {
```

File: backend/src/routes/api/components/list.ts [hunk 17]
```diff
@@ -36,30 +36,33 @@ export const removeComponent = async (
...
     };
-    await coreV1Api.replaceNamespacedConfigMap(enabledAppsConfigMapName, namespace, cmBody);
+    await coreV1Api.replaceNamespacedConfigMap(enabledAppsConfigMapName ?? '', namespace, cmBody);
     await updateApplications();
-    return { success: true, error: null };
+    return { success: true, error: '' };
   } catch (e) {
-    fastify.log.error(e.message);
-    return { success: false, error: e.message };
+    const message = e instanceof Error ? e.message : '... some default error message ....';
```
Comment (christianvogt) [inline backend/src/routes/api/components/list.ts:64]:
```suggestion
    const message = e instanceof Error ? e.message : 'Error removing component.';
```

File: backend/src/types.ts [hunk 18]
```diff
@@ -148,6 +148,18 @@ export type K8sResourceCommon = {
...
+  metadata: {
+    name?: string;
+    namespace: string;
+    generateName?: string;
+    uid?: string;
+    labels?: { [key: string]: string };
+    annotations?: { [key: string]: string };
+    creationTimestamp?: Date;
+  };
+} & K8sResourceBase;
```
Comment (christianvogt) [inline backend/src/types.ts:161]: Does this work? (also made the type name smaller)
```suggestion
export type K8sNamespacedResourceCommon = {
  metadata: {
    namespace: string;
  };
} & K8sResourceCommon;
```

File: backend/src/utils/route-security.ts [hunk 19]
```diff
@@ -63,17 +63,17 @@ const requestSecurityGuardNotebook = async (
...
+  name?: string,
 ): Promise<void> => {
   const { notebookNamespace, dashboardNamespace } = getNamespaces(fastify);
   const username = await getUserName(fastify, request);
   const translatedUsername = usernameTranslate(username);
   const isReadRequest = request.method.toLowerCase() === 'get';
 
   // Check to see if a request was made against one of our namespaces
-  if (![notebookNamespace, dashboardNamespace, MODEL_REGISTRY_NAMESPACE].includes(namespace)) {
+  if (![notebookNamespace, dashboardNamespace].includes(namespace)) {
```
Comment (christianvogt) [inline backend/src/utils/route-security.ts:76]: Why is `MODEL_REGISTRY_NAMESPACE` removed?

Comment (rsun19) [reply backend/src/utils/route-security.ts]: I have no idea how that happened

File: backend/src/utils/route-security.ts [hunk 20]
```diff
@@ -161,23 +179,22 @@ const handleSecurityOnRouteData = async (
...
       request.params.namespace,
       needsAdmin,
+      request.params.name,
     );
   } else if (isRequestNotebookAdmin(request)) {
-    await requestSecurityGuardNotebook(fastify, request, request.body.username);
+    await requestSecurityGuardNotebook(fastify, request, request.body.username ?? '');
   } else if (isRequestNotebookEndpoint(request)) {
     // Endpoint has self validation internal
-    return;
```
Comment (christianvogt) [inline backend/src/utils/route-security.ts:180]: Undo change.

**Comment by christianvogt** [comment]
/lgtm
/approve

did a bunch of ad hoc testing of various parts of the UI
